finality-aleph/src/sync/substrate/verification/cache.rs

```diff
@@ -128,9 +156,68 @@ where
     /// Prune all sessions with a number smaller than `session_id`
     fn prune(&mut self, session_id: SessionId) {
         self.sessions.retain(|&id, _| id >= session_id);
+        // off by one!
+        // let's retain one more session, just to be safe
+        self.aura_authorities
+            .retain(|&id, _| id.next() >= session_id);
         self.lower_bound = session_id;
     }
 
+    fn try_prune(&mut self, session_id: &SessionId) {
+        if session_id.0
+            >= self
+                .lower_bound
+                .0
+                .saturating_add(self.cache_size.saturated_into())
+        {
+            self.prune(SessionId(
+                session_id
+                    .0
+                    .saturating_sub(self.cache_size.saturated_into())
+                    + 1,
+            ));
+        }
+    }
+
+    /// Returns the list of Aura authorities for a given block number. Updates cache if necessary.
+    /// This method assumes that the queued Aura authorities will indeed become Aura authorities
+    /// in the next session.
+    pub fn get_aura_authorities(
+        &mut self,
+        number: BlockNumber,
+    ) -> Result<&Vec<AuraId>, CacheError> {
+        if number == 0 {
+            return Err(CacheError::NoAuraAutoritiesForGenesis);
```

Comment:
Shouldnt we check at least that header at zero match genesis header?

Reply:
Also, we pass `parent_number` to this function, which means we will actually run this code for block number one, which should have authorities – is this a bug or am I missing something?

Reply:
Tbh I would leave the call as is and remove this error, while getting the session of the next block below – otherwise this will be an error that never gets returned as we never call this method for 0, right?

Reply:
it's checked in `VerifierCache::verify_header`, here we only have access to the id
currently this error will never be returned

Reply:
I think I've fixed it, but please take a look at the full commit just to be sure